### PR TITLE
Add patches for Kustomize to add objectSelectors to pod webhook configurations

### LINKED ISSUE #361: [JobSet v0.3.0 Bug] Webhook failing to start and somehow blocking regular Jobs not owned by JobSets from creating pods
This problem doesn't happen consistently, as I manually tested JobSet v0.3.0 multiple times successfully. However, a user today is running into a concerning error. After installing v0.3.0, the JobSet deployment manager fails to start up, then somehow when they submit a regular Job (not owned by a JobSet), the Job cannot create any pods due to an error calling the JobSet webhook (??).

**Steps to reproduce:**

1. Deploy JobSet v0.3.0

Jobset controller manager not starting properly, no endpoints for webhook service:

`kubectl describe rs -n jobset-system`
```
...
Events:
  Type     Reason        Age                  From                   Message
  ----     ------        ----                 ----                   -------
  Warning  FailedCreate  8m57s (x47 over 8h)  replicaset-controller  Error creating: Internal error occurred: failed calling webhook "mpod.kb.io": failed to call webhook: Post "https://jobset-webhook-service.jobset-system.svc:443/mutate--v1-pod?timeout=10s": no endpoints available for service "jobset-webhook-service"
```

2. Create a standalone Job.  Example Job + headless service spec being deployed by the user encountering this issue:

```
apiVersion: v1
kind: Service
metadata:
  name: headless-svc
spec:
  clusterIP: None
  selector:
    job-name: tpu-job
---
apiVersion: batch/v1
kind: Job
metadata:
  name: tpu-job
spec:
  backoffLimit: 0
  # Completions and parallelism should be the number of chips divided by 4.
  # (e.g. 4 for a v5litepod-16)
  completions: 4
  parallelism: 4
  completionMode: Indexed
  template:
    spec:
      subdomain: headless-svc
      restartPolicy: Never
      terminationGracePeriodSeconds: 300
      nodeSelector:
        cloud.google.com/gke-tpu-accelerator: tpu-v4-podslice
        cloud.google.com/gke-tpu-topology: 2x2x1
      containers:
      - name: tpu-job
        image: python:3.10
        ports:
        - containerPort: 8471 # Default port using which TPU VMs communicate
        - containerPort: 8431 # Port to export TPU usage metrics, if supported
        securityContext:
          privileged: true
        command:
        - bash
        - -c
        - |
          printenv
          pip install "jax[tpu]" -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
          python -c 'import jax; print("JAX Devices:", jax.devices(), "Global device count:", jax.device_count(), "Local device count:", jax.local_device_count())'
        resources:
          requests:
            google.com/tpu: 4
          limits:
            google.com/tpu: 4
```

Describing the Job after deploying it:

`kubectl describe job tpu-job`

```

Name:               tpu-job
Namespace:          default
Selector:           batch.kubernetes.io/controller-uid=d0adcb30-8b41-4122-9751-2bb502f845d9
Labels:             batch.kubernetes.io/controller-uid=d0adcb30-8b41-4122-9751-2bb502f845d9
                    batch.kubernetes.io/job-name=tpu-job
                    controller-uid=d0adcb30-8b41-4122-9751-2bb502f845d9
                    job-name=tpu-job
Annotations:        <none>
Parallelism:        4
Completions:        4
Completion Mode:    Indexed
Start Time:         Thu, 21 Dec 2023 17:31:55 +0000
Pods Statuses:      0 Active (0 Ready) / 0 Succeeded / 0 Failed
Completed Indexes:  <none>
Pod Template:
  Labels:  batch.kubernetes.io/controller-uid=d0adcb30-8b41-4122-9751-2bb502f845d9
           batch.kubernetes.io/job-name=tpu-job
           controller-uid=d0adcb30-8b41-4122-9751-2bb502f845d9
           job-name=tpu-job
  Containers:
   tpu-job:
    Image:       python:3.10
    Ports:       8471/TCP, 8431/TCP
    Host Ports:  0/TCP, 0/TCP
    Command:
      bash
      -c
      printenv
      pip install "jax[tpu]" -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
      python -c 'import jax; print("JAX Devices:", jax.devices(), "Global device count:", jax.device_count(), "Local device count:", jax.local_device_count())'
      
    Limits:
      google.com/tpu:  4
    Requests:
      google.com/tpu:  4
    Environment:
      TPU_WORKER_HOSTNAMES:  tpu-job-0.headless-svc,tpu-job-1.headless-svc,tpu-job-2.headless-svc,tpu-job-3.headless-svc
      TPU_WORKER_ID:          (v1:metadata.annotations['batch.kubernetes.io/job-completion-index'])
    Mounts:                  <none>
  Volumes:                   <none>
Events:
  Type     Reason        Age                From            Message
  ----     ------        ----               ----            -------
  Warning  FailedCreate  30s (x6 over 61s)  job-controller  Error creating: Internal error occurred: failed calling webhook "mpod.kb.io": failed to call webhook: Post "https://jobset-webhook-service.jobset-system.svc:443/mutate--v1-pod?timeout=10s": no endpoints available for service "jobset-webhook-service"
$:~/go/src/sigs.k8s.io/jobset$ k get pods -n jobset-system
```

### PR DESCRIPTION
**Pull request by danielvegamyhre**
Fixes #361 

Note I originally tried using the `add` JSON Patch op type to simply add objectSelector field into the existing webhook manifests, but it seems to add the objectSelector as a new element in the `webhooks` array, rather than adding the field to the element at index 1 in the array.

```
- objectSelector:
       matchExpression: ...
```
instead of:
```
  objectSelector:
       matchExpression: ...
```

The only way I could get it to work was by completely replacing it. This isn't ideal since manifest changes done via kubebuilder markers in the Go code will just get replaced, so we'd need to update the patch every time we update the markers. However, I don't anticipate us needing to update that marker frequently (or ever, really).

References:

- https://github.com/kubernetes-sigs/kustomize/blob/master/examples/jsonpatch.md
- https://jsonpatch.com/#add
- https://datatracker.ietf.org/doc/html/rfc6902

### CODE REVIEW

File: config/components/webhook/mutating-patch.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+- op: replace
```
Comment (ahg-g) [inline config/components/webhook/mutating-patch.yaml]: does this override the the configuration for the jobset webhooks? the one named "mutate-jobset-x-k8s-io-v1alpha2-jobset"?

Comment (danielvegamyhre) [reply config/components/webhook/mutating-patch.yaml]: Nope, just the mutating webhook configuration at index 1 in the mutating-webhook-configuration `webhooks` array, which is the pod mutating webhook: https://github.com/kubernetes-sigs/jobset/blob/dbabc65129107fd830e7f6b367e9223c87677d7e/config/components/webhook/manifests.yaml#L27-L44

Same goes for the validating-webhook-configuration patch.

**Comment by ahg-g** [comment]
Thanks @danielvegamyhre; we should create a minor release with this important fix.

/lgtm

**Comment by ahg-g** [comment]
/lgtm
/hold

lift the hold once you feel comfortable merging

**Comment by danielvegamyhre** [comment]
Ok I manually tested these changes by installing JobSet on a live cluster which has nodes with insufficient CPU resources to run the controller manager deployment pod, meaning the webhook configuration objects are installed on the cluster, but the deployment pod is unschedulable / in a pending state. 

This effectively reproduces the scenario in #361 and the expected behavior is that other pods in the cluster are NOT blocked from being created, since Jobset's pod webhook should not be selecting those pods, and indeed that is the behavior we see with these changes.

**Pull request merged**